### PR TITLE
Multiprocessing on cpu optimizer for sync scheduling

### PR DESCRIPTION
**Pull request by ttochi**
### Features
- Divide optimizer for each bwd stage's parameter (if there are 3 bwd stage, 3 optimizer will be created)
- A new `SpiralOptimizer` class will manage these optimizer list
- Make optimizer.step to run after each bwd stage has offloaded gradients
- Make optimizer.step to run as subprocess to overlap with main process

### Comments
In cpu process, we need to use `cudaEventSynchronize()` instead of `cudaStreamWaitEvent()`.
So I created `sync_event` method in SpiralCUDAManager.

### CODE REVIEW

**Review by gajagajago — APPROVED** [review]
Thanks for the PR. I think this PR is very brilliant.

Approval, but I want a misc code modification.
Please remove `torch.cuda.nvtx.range_push(f"opt fork:{bwd_stage_id}")` and `torch.cuda.nvtx.range_push(f"opt join")`, and just annotate the weight_update function. It is okay if we see multiple processes with the annotation on nsys report.

Also, I think it would be good to have a test code for spiral cpu optimizer, that shows the correctness of our "multiple distributed(?)" optimizer.

**Comment by ttochi** [comment]
> Please remove torch.cuda.nvtx.range_push(f"opt fork:{bwd_stage_id}") and torch.cuda.nvtx.range_push(f"opt join"), and just annotate the weight_update function.

Actually, I tried to attach nvtx tag on subprocess logic, but it was hard to track forked subprocess on nsys.
I found that we need to `spawn` process instead of `fork`  to track subprocess in nsys. ([reference](https://stackoverflow.com/questions/74748415/nvtx-markers-with-python-multiprocessing))

We can set `spawn` method in multiprocessing library. The `spawn` method use pickle to serialize functions. However, Megatron code uses a lot of local functions that are not serializable with pickle.

I tried some workarounds in following [reference ](https://stackoverflow.com/questions/74748415/nvtx-markers-with-python-multiprocessing) but fail to track nvtx😭

**Comment by ttochi** [comment]
> Also, I think in a following PR it would be good to have a test code for spiral cpu optimizer, that shows the correctness of our "multiple distributed(?)" optimizer.

Good suggestion! I will add unit test code :)

**Comment by gajagajago** [comment]
Got it. Thx!
